cura/Settings/CuraContainerStack.py

```diff
@@ -0,0 +1,616 @@
+#   - quality changes: An InstanceContainer with the metadata entry "type" set to "quality_changes".
+#   - quality: An InstanceContainer with the metadata entry "type" set to "quality".
+#   - material: An InstanceContainer with the metadata entry "type" set to "material".
+#   - variant: An InstanceContainer with the metadata entry "type" set to "variant".
+#   - definition changes: An InstanceContainer with the metadata entry "type" set to "definition_changes".
+#   - definition: A DefinitionContainer.
+#
+#   Internally, this class ensures the mentioned containers are always there and kept in a specific order.
+#   This also means that operations on the stack that modifies the container ordering is prohibited and
+#   will raise an exception.
+class CuraContainerStack(ContainerStack):
+    def __init__(self, container_id: str, *args, **kwargs):
+        super().__init__(container_id, *args, **kwargs)
+
+        self._empty_instance_container = ContainerRegistry.getInstance().getEmptyInstanceContainer()
+
+        self._containers = [self._empty_instance_container for i in range(len(_ContainerIndexes.IndexTypeMap))]
+
+        self.containersChanged.connect(self._onContainersChanged)
+
+    # This is emitted whenever the containersChanged signal from the ContainerStack base class is emitted.
+    pyqtContainersChanged = pyqtSignal()
+
+    ##  Set the user changes container.
+    #
+    #   \param new_user_changes The new user changes container. It is expected to have a "type" metadata entry with the value "user".
+    def setUserChanges(self, new_user_changes: InstanceContainer) -> None:
+        self.replaceContainer(_ContainerIndexes.UserChanges, new_user_changes)
+
+    ##  Get the user changes container.
+    #
+    #   \return The user changes container. Should always be a valid container, but can be equal to the empty InstanceContainer.
+    @pyqtProperty(InstanceContainer, fset = setUserChanges, notify = pyqtContainersChanged)
+    def userChanges(self) -> InstanceContainer:
+        return self._containers[_ContainerIndexes.UserChanges]
+
+    ##  Set the quality changes container.
+    #
+    #   \param new_quality_changes The new quality changes container. It is expected to have a "type" metadata entry with the value "quality_changes".
+    def setQualityChanges(self, new_quality_changes: InstanceContainer) -> None:
+        self.replaceContainer(_ContainerIndexes.QualityChanges, new_quality_changes)
+
+    ##  Set the quality changes container by an ID.
+    #
+    #   This will search for the specified container and set it. If no container was found, an error will be raised.
+    #
+    #   \param new_quality_changes_id The ID of the new quality changes container.
+    #
+    #   \throws Exceptions.InvalidContainerError Raised when no container could be found with the specified ID.
+    def setQualityChangesById(self, new_quality_changes_id: str) -> None:
```

Comment:
Nevermind. The code already does this. The replaceContainer checks if the type is correct (and setQualityChanges uses replace container)